feat(blink): improve call quality

sdwoodbury opened the pull request:
<!--  Thanks for sending a pull request!-->

### What this PR does 📖

- use the latest version of Warp, which includes a breaking update to Blink implementation. It listens on different gossipsub channels, meaning that anyone using the old software won't be able to participate in calls with peers who are using newer versions of the software. 
- update logging so that blink logs are included in debug logs, and added an option to `trace-blink`. 
- updated the way call state (muted/unmuted/etc) is tracked for participants. 

### Which issue(s) this PR fixes 🔨

- Resolve #1409 
- Resolve #1474 

<!--Add the ticket Github number such as #Resolve #001 to automatically link the PR to the issue-->

### Special notes for reviewers 🗒️

### Additional comments 🎤
wait for https://github.com/Satellite-im/Uplink/pull/1569 to be merged first

sdwoodbury committed:
bump blink - improve signaling

sdwoodbury committed:
Merge branch 'dev' into fix/blink-signaling

sdwoodbury committed:
bump warp

sdwoodbury committed:
Merge branch 'dev' into fix/blink-signaling

sdwoodbury committed:
bump warp version

sdwoodbury committed:
chore(call): follow the design pattern

sdwoodbury committed:
fix clippy

dariusc93 committed:
Merge branch 'dev' into fix/blink-signaling

sdwoodbury committed:
improve logging

sdwoodbury committed:
chore(logging): change some debug! to trace!

sdwoodbury committed:
chore(logging): change more debug! logs to trace!

sdwoodbury committed:
move more logs to trace

sdwoodbury committed:
bump warp version

sdwoodbury committed:
Merge branch 'fix/blink-signaling' of github.com:Satellite-im/Uplink into fix/blink-signaling

sdwoodbury committed:
update uplink to reflect changes in blink

sdwoodbury committed:
Merge branch 'dev' into fix/blink-signaling

sdwoodbury committed:
update dependency

sdwoodbury committed:
allow blink to run without a mp4 logger for now

sdwoodbury committed:
bump warp rev

sdwoodbury committed:
bump warp rev

sdwoodbury committed:
bump warp revision

sdwoodbury committed:
bump warp

sdwoodbury committed:
bump warp. fix mute and echo cancellation

sdwoodbury committed:
bump warp. try to fix mute for real this time

sdwoodbury committed:
merge with dev and bump warp

sdwoodbury committed:
Merge branch 'dev' into fix/blink-signaling

sdwoodbury committed:
bump warp revision

sdwoodbury committed:
make utility function to get identities from call

lgmarchi commented on common/src/state/call.rs:
```diff
@@ -113,36 +112,36 @@ impl CallInfo {
     }
 
-    pub fn participant_left(&mut self, call_id: Uuid, id: &DID) -> anyhow::Result<()> {
+    pub fn participant_joined(&mut self, call_id: Uuid, id: DID) -> anyhow::Result<()> {
```
Stu, following clean code, could we change to handle_participant_speaking? Functions are supposed to start with a verb. Same thing for other functions here.

dariusc93 replied:
Generally, it is usually more common to have the function contain a verb than to start with one when they are performing a action that modifies data or context (in this case, the function is mutable with it showing "joined"). However,  that usually comes down to the overall coding style.

sdwoodbury replied:
@lgmarchi  I think the verb "handle" is too generic to be useful. Also, this function updates state when someone joins a call, not when they speak, so "speaking" shouldn't be in the function name either. I respectfully reject this proposal.

lgmarchi commented on ui/src/bootstrap.rs:
```diff
@@ -89,6 +89,7 @@ pub fn configure_logger(profile: Option<LogProfile>) {
     let max_log_level = if let Some(profile) = profile {
         match profile {
             LogProfile::Debug => {
+                logger::allow_other_crates(Level::Debug, Some(&["warp-blink-wrtc"]));
```
Could we let "warp-blink-wrtc", as const variable in the page and repeat it in the other places?

lgmarchi commented on ui/src/logger/mod.rs:
```diff
@@ -56,10 +56,8 @@ pub struct Logger {
 
-    // display trace logs from uplink
-    uplink_trace: bool,
     // minimum log level for 3rd party crates (warp, dioxus, etc)
```
Need to change the comment, to be equal variable context.

sdwoodbury replied:
@lgmarchi I don't understand what comment needs to be changed and why. perhaps bad grammar is causing some confusion.

lgmarchi replied:
You changed the name of variable for max_level, and comment is as minimum log level for 3rd party

stavares843 replied:
marking this as resolved to get this merged but in case you wanna address @lgmarchi comment above cc @sdwoodbury

lgmarchi reviewed (approved):
Left some comments to improve code quality.

zeenix commented on common/src/lib.rs:
```diff
@@ -48,6 +48,8 @@ pub enum LogProfile {
     TraceAll,
+    /// maximum logging for blink
+    TraceBlink,
 }
```
You're just adding another log option here so this comment is more generic: typically, people just make use of the `RUST_LOG` env, where you can set different log levels for different crates (and even mods). Was there a reason for going with custom/specific method?

dariusc93 replied:
To my knowledge, the custom method was done so it would allow routing the logs to the component in uplink vs just outputting it to a file or to terminal

zeenix replied:
hmm.. I'm not an expert on logging but aren't the loglevel control and output routing decoupled in Rust logging/tracing crates?

dariusc93 replied:
To my knowledge they are, but these are more specific to the [logging component](https://github.com/Satellite-im/Uplink/blob/dev/ui/src/components/debug_logger/mod.rs#L13) in uplink.

zeenix replied:
> To my knowledge they are, but these are more specific to the [logging component](https://github.com/Satellite-im/Uplink/blob/dev/ui/src/components/debug_logger/mod.rs#L13) in uplink.

That's the question: why are they specific to Uplink? Sorry, I wasn't clear.

sdwoodbury replied:
@zeenix  is there a way to make the logger emit the trace logs for one crate and ignore the trace logs for all other crates using environment variables? So far I only know how to use `RUST_LOG` to set the same log level for all crates.

zeenix replied:
> @zeenix is there a way to make the logger emit the trace logs for one crate and ignore the trace logs for all other crates using environment variables?

Yes, sir. As I mentioned, you can even set different levels for mods. E.g `RUST_LOG=tokio=info,uplink=trace,zbus=debug,zbus::connection=trace` sets different levels for different crates and then sets a different level for `connection` mod of `zbus` than rest of `zbus`.

sdwoodbury replied:
@zeenix  this does look a lot better. we could probably get rid of most of the logging code and just use RUST_LOG. maybe it's better to do that in a separate PR.

zeenix replied:
> maybe it's better to do that in a separate PR.

Definitely a different PR. :)

zeenix commented on common/src/state/call.rs:
```diff
@@ -37,9 +37,8 @@ pub struct Call {
-    pub participants_joined: Vec<DID>,
+    pub participants_joined: HashMap<DID, ParticipantState>,
     pub participants_speaking: HashMap<DID, Instant>,
-    pub participants_state: HashMap<DID, ParticipantState>,
```
You did a great job with PR description (unlike me :laughing:) so I hate to ask this but it would be good to summarize the specific changes like this one so people new to the code (like me) can more easily understand what's going on. It will also help future archeologist (which may be you :smile:).

sdwoodbury replied:
@zeenix I combined lines 40 and 42. I thought it unnecessary to maintain a vec of participants who joined and also a `HashMap<DID, ParticipantState>` because that map only contains state for participants who have joined the call.

zeenix replied:
@sdwoodbury Thanks! Now please also help the future archeologists. :)

zeenix commented on common/src/state/mod.rs:
```diff
@@ -335,9 +336,18 @@ impl State {
+        ) {
+            {
+                log::trace!("process_warp_event: {event}");
+            }
```
why the extra `{}`? :thinking:

sdwoodbury replied:
i'll remove it

zeenix commented on common/src/state/mod.rs:
```diff
@@ -720,14 +730,14 @@ impl State {
                 }
             }
+            BlinkEventKind::AudioStreamError => {
+                // todo
```
might be good to log a debug or trace here.

zeenix commented on ui/src/components/media/calling.rs:
```diff
@@ -101,7 +101,7 @@ pub struct ActiveCallProps {
 fn ActiveCallControl(cx: Scope<ActiveCallProps>) -> Element {
```
unrelated to your PR but this should be `activate_call_control`.

dariusc93 replied:
This was done to match the same style as dioxus macro components

zeenix replied:
interesting, do you have a link to some example?

dariusc93 replied:
If you mean within uplink, then the function [CallUserImageGroup](https://github.com/Satellite-im/Uplink/blob/a7c216be2bf0948fd14ca7688d7c6798caf0431b/ui/src/components/media/calling.rs#L679) is used [here](https://github.com/Satellite-im/Uplink/blob/a7c216be2bf0948fd14ca7688d7c6798caf0431b/ui/src/components/media/calling.rs#L383-L385). In dioxus, you can probably see it more in some of their examples (eg https://github.com/DioxusLabs/dioxus/blob/b182a6ee75c3f69d5498de3b6d8122b94c8c8fbc/examples/borrowed.rs)

zeenix replied:
Thanks, that's super strange. I've never seen any Rust project make an exception to this convention.

sdwoodbury committed:
Merge branch 'dev' into fix/blink-signaling

sdwoodbury committed:
chore(logging): replace string with a const when configuring logging

sdwoodbury committed:
chore(logging): add log statement and clean up code

sdwoodbury committed:
update warp again
accidentally reverted warp during the merge.
updated to the latest version of warp due to a bugfix, so the current hash of warp won't match that from before the merge

sdwoodbury committed:
Merge branch 'dev' into fix/blink-signaling

phillsatellite commented:
Wanted to leave a comment so everyone knew where we were at. @tooshel @stavares843 and I tested today while doing a group call. It sounded very clear and there was no gargling noise but we did notice that there was some major delay during the conversation. I would say something to Sheldon and about 5 seconds later I would hear what I said through Sheldons speakers. There were other times when Sheldons sentence would get cut-off and I would just miss half of what he was saying. Nonetheless this is still working better than dev in my opinion and would probably be worth merging in and making further improvements in new tickets. Tomorrow morning @tooshel @stavares843 and I were going to do another call while recording it and share it here so we can better explain the issues that we faced. 

Added changes requested till Sheldon/Sara and I talk tomorrow morning to see if we want to merge it in

sdwoodbury committed:
Merge branch 'dev' into fix/blink-signaling
